### PR TITLE
Plando through a "distribution file".

### PR DESCRIPTION
**Pull request by Exter-N**
The "distribution file" is an additional JSON configuration file that provides a way to partially or completely bypass the randomness and :
- Customize the hash (the five icons on top of the menu) ;
- Anchor specific items (or random junk) to specific locations ;
- Anchor specific stones/medallions to specific bosses ;
- Choose/skip specific trials ;
- Pick vanilla/MQ versions of specific dungeons ;
- Place some keys out of their dungeon, without fully enabling keysanity ;
- Disguise all freestanding items, not just the Ice Traps ;
- Customize what Gossip Stones say ;
- Replace items in the pool (specific with specific, random junk with specific, specific with random junk) ;
- It can recreate vanilla behavior (#89 #229) ;
- It can restore the freestanding HP/HC/SK/GS models (#377 #390).

Additionally, when "Create Spoiler Log" is checked, a second copy of the spoiler log is generated in the "distribution file" format, to allow for easier tweaking of the seed.

Example of a "distribution file" (created by hand) :
```json
{
    "file_hash": [ "Longshot", "Longshot", "Longshot", "Longshot", "Longshot" ],
    "dungeons": [ {
        "Water Temple": { "mq": true },
        "Spirit Temple": { "mq": false }
    } ],
    "trials": [ {
        "Light": { "skip": false },
        "Water": { "skip": true }
    } ],
    "item_replacements": [ [
        { "remove": "Heart Container", "add": "Piece of Heart" },
        { "remove": "#Junk", "add": "Piece of Heart", "count": 3 }
    ] ],
    "locations": [ {
        "Queen Gohma": { "item": "Kokiri Emerald" },
        "Links Pocket": { "item": "*Medallion" },
        "Gerudo Training Grounds Underwater Silver Rupee Chest": { "item": "Rupees (5)" },
        "Mido Chest*": [
            { "item": "Double Defense" },
            { "item": "Boss Key (Spirit Temple)" }
        ],
        "Lon Lon Tower Freestanding PoH": { "model": "Rupee (1)" }
    } ],
    "gossip": [ {
        "Kokiri Forest (Storms)": { "gossip": "We wish you a merry Christmas!" }
    } ]
}
```

### CODE REVIEW

**Comment by Exter-N** [comment]
By the way, a distribution file could (certainly) be used to recreate vanilla behavior (#89 #229).

**Comment by TestRunnerSRL** [comment]
In general I really love that you're doing this. I had planned on trying to get something like this done, so big kudos on working on it. That being said there's some things I'll probably plan on changing. I'm unsure if this is something I'll try to get in before 4.0 but I'll try. It's a bit less critical since this doesn't really affect the webpatcher and that's the only thing that cares about 4.0. What I'll probably do is merge this into a new branch.

1) We need to salt the seed in some way. Like hash the entire file and salt the seed with that.
2) I'm not really a big fan of having two competing spoiler formats. I think having it in a json is a good idea, but we should work on deprecating the old on and having the spoiler just output in a compatible json.
3) I feel like the new spoiler is a bit wordy and the multiworld structure is a little odd. Like for example instead of doing 
```
    "trials": [
        { "trial": "Light", "skip": false },
        { "trial": "Water", "skip": true }
    ],
```
I would do
```
    "trials": {
        "Light": false,
        "Water": true
    },
```
I think `anchors` should be named `locations` instead. `boss_anchors` could also probably be combined with `locations` as well. Having some kind of alignment in the json like the spoiler has would be nice but I'm probably gonna bother asking for that since we're just doing a json dump which I don't believe has any functionality for.
4) When applying, I think is kinda odd how there's so much single/multiworld code when I think most of it can be unified.
5) I don't think we should allow setting the item look-alike if we aren't setting the item. I also don't really like this feature much, and I'm pretty hesitant about it. I kinda like that it allows some major trolling in plando, but I think people would honestly just be really annoyed by that, since that's pretty far from normal rando behavior. There's also a reason those 2 other PR have not been merged.
6) I think junk achors should probably should have the item explicitly set to "Junk"
7) I don't like the error reporting, it's pretty vague and almost condescending. Instead of saying `'What is a "%s"'` it should be `'Unknown Item %s'`
8) I think most of this stuff should probably be moved out of Fill.py and into ItemPool,py
9) I haven't tested this too thoroughly, but be careful about the gossip stones. The text in the spoiler is not the correct text that should be used and has a lot of important info stripped. So when loading in it needs to add in the various line breaks (which I think is being handled), but it's completely lacking the the colors. So we'll need to add in a way to colorize the text properly again, or leave the hex command codes in.
10) We should allow adding items that are not in the pool. What I had planned for this was that it would try to remove the item from the pool, and if it failed then it would remove a random junk item. If there's no remaining junk items then I'm not sure. Maybe it can fail at that point, or maybe it would start removing non junk items. I think we should allow a plando that has literally nothing but the hookshot if they so choose.
11) Not really a change but keep in mind that plando will often create seeds that are unbeatable by logic but are actually still beatable. This is the primary reason I added a no-logic mode.
12) I'm unsure about the `replacement` system. Seems kinda hard to use, and a lot less important if the item doesn't need to be in the pool to place. I feel like it has a strong use but I just am unsure about the current implementation. Maybe it would be better to just list out the whole item pool?
13) You really just allow some item placements without certain settings. Things like song shuffle for example make major changes to the code to allow it to be possible, so you really can't do anything without that option on. These restrictions are enforced primarily by the `location.can_fill_fast()`, which is being ignored. I think that we should still make sure that the item item is placable. I think the main thing you're trying to get around is keysanity which is actually doable. I'm kinda unsure if it's something we should allow tho in anyways.

There's probably more stuff but this is my initial thoughts going over it. this will likely need multiple revisions.

**Comment by Exter-N** [comment]
And 13 is fixed.

**Comment by Exter-N** [comment]
More about 3 : the current format allows to apply some instructions to all the worlds in a multiworld setting, but some other instructions to one world only. That would be harder to represent with a format like `"trials":{"Light":false,"Water":true}` (plus there is the usual problem with booleans : does `true`/`false` means skip/don't skip, or keep/don't keep, or do something else/don't do something else?)

**Comment by TestRunnerSRL** [comment]
1. You need to only load the file if it exists, it's currently always trying to load it. Also if there's a problem opening the file then this will crash at the top level, so it's completely unhandled.
2. I think JSON is fairly readable, and it would be actually pretty trivial to add the missing info. If you want a reference, plenty of other randomizers already do this, like Metroid Prime and Wind Waker.
3. If there's a concern about booleans, just make them strings with stuff like "Skipped" or wtv. If you need more properties for a single entry, then make the value be a object instead of an primitive. so like 
```
"locations": {
    "Gerudo Training Grounds Underwater Silver Rupee Chest": {"item": "Rupees (5)"}
}
```
5. I don't have any problems with people making forks with their own special modifications. I don't really think it's equatable to Ice Traps. With Ice Traps if you check something you think is important but isn't you just waste some time. But the other way around where you skip something that looks useless, then it will make you stuck completely, and that's really shitty. I realize this falls under the "This is fine if you use your powers only for good", but let's be real here, everyone is going to try to be an jerk.
6. Don't make any plans that a GUI will exist for this. I wouldn't even count on the GUI being properly maintained for the normal stuff since everything is moving to the webpatcher. I say this because I think this opens up the potential of making item groups. Like "junk" is really just some set of item types, but there's no reason you shouldn't be able to have a group for "major item" and so forth.
10. I dunno, I just find filling something like this out incredibly tedious if I'm making major changes. I should only care about the itempool for things that I'm not explicitly placing. Like how it should work is that after the items are placed down from this file it will then start replacing the remaining pool with this list. I'm not really sure you're quite aware how much a problem this is because the number of items in the item pool isn't even known depending on your settings. So like if I place every location then my item pool should just be empty. I think having the replace typing for this is kinda reasonable if you just want to change the item pool without doing a plando, but this structure makes it very difficult otherwise. I think this needs to be a lot more flexible. Also the naming is really confusing since `replacements` has no indication it's talking about the item pool. If I had a suggestion it would be to instead maybe be able to list how many items you want in the pool. Like you could say `"nayru's love": 1` and if there was one missing it would replace a random item (starting with junk) with it. If there's extra then it would replace them with random junk.
13. I suppose but you could also just turn keysanity on. A plando will probably be placing every key by hand anyways.

**Comment by TestRunnerSRL** [comment]
Also didn't mention about the formatting the multiworld, it's pretty unusual to want to specify some property for all worlds. I think it would make more sense to not have a world property at all and to put in a list of entries (one for each world) like it is structured in the spoiler.

**Comment by Exter-N** [comment]
1. I fixed that one recently, now if a path pointing to a nonexistent file is specified, it's ignored ;
2. I'm testing the feature with a few users that aren't devs and they have trouble working with JSON, they could be taught, but for novices a viewer/editor would still probably be a better option (more about that in 6) ;
3. I'm thinking about a structure like this (`_comment` keys are just comments for here and aren't part of the actual structure), what do you think about it? As I already said before, it would apply to both input (distribution file) and, where applicable, output (JSON spoiler).
```json
{
    "file_hash": [ "as it is currently" ],
    "dungeons": [
        {
            "_comment": "Dungeon configuration for first world/player",
            "Water Temple": { "mq": true, "_comment": "Easier to parse than a string, allows for future extension, and still no unlabeled booleans" },
            "Spirit Temple": { "mq": false }
        },
        {
            "_comment": "Dungeon configuration for second world/player"
        }
    ],
    "trials": [
        {
            "_comment": "Trial configuration for first world/player",
            "Light": { "skip": false },
            "Water": { "skip": true }
        }
    ],
    "_comment:item_pools": "Like all the keys in the file, this one is optional. If specified, it completely replaces the item pools of the worlds where its value is not null.",
    "item_pools": [
        {
            "Heart Container": { "count": 8 },
            "Piece of Heart": { "count": 3 },
            "Bottle with Letter": { "_comment": "count is implicitly 1 if omitted" },
            "#Bottle": { "count": 3, "_comment": "3 bottles with random contents, other than the note that mweeps – # denotes a group" }
        },
        {
            "Progressive Hookshot": { "count": 2, "_comment": "the pool will be padded with random junk" }
        }
    ],
    "_comment:item_replacements": "Still useful if you want to start from a standard pool. Usefulness in combination with item_pools is limited, but there might still be some in plando/rando hybrids, for example, having 4 bottles with different fixed contents, then replacing a random one with the fishy stationery.",
    "item_replacements": [
        [
            { "old_item": "#Junk", "new_item": "Rupees (5)", "count": 2, "_comment": "Force two junk items to be blue rupees for first world/player" }
        ],
        [
            { "old_item": "Progressive Hookshot", "new_item": "#Junk", "_comment": "count is implicitly 1 if omitted, so, actually, no longshot for second world/player" }
        ],
    ],
    "locations": [
        {
            "Morpha": { "item": "!*Medallion", "_comment": "Make Morpha optional. Well, unless we don't make sure it's not going to be an all-dungeons ROM and the generator decides that it will be one. Or Morpha Heart ends up on the way of the hero." },
            "Morpha Heart": { "item": "Iron Boots", "extra": true, "_comment": "They say that it's a bad idea to put these here. Fortunately, they are going to consume a random junk item from the pool and let the original IB go somewhere else." },
            "Zoras Fountain Bottom Freestanding PoH": { "item": "Light Arrows", "model": "Rupees (5)" },
            "Mido Chest*": [
                { "item": "Double Defense", "player": 1, "_comment": "Single name as key => single record as value, pattern as key => array of records as value" },
                { "item": "Double Defense", "_comment": "If player is omitted, item is taken from the same world as the location it's put in" }
            ]
        }
    ],
    "gossip": [
        {
            "Kokiri Forest (Storms)": { "gossip": "They say I'm on the way of the hero.", "_comment": "Hopefully you found the Song of Storms before the Stone of Agony. Or you have a good memory. Or you take notes." },
            "Zoras Domain (Mweep)": { "gossip": "They say a great treasure behind me is wearing a Stone Mask." }
        }
    ]
}
```
4. Indeed, seems like GitHub renumbered everything :trollface: ;
5. If you really don't want it I'll remove it from the PR, but it's certainly easy enough for people who want to do something creative or just really want to be jerks to add it back that it won't stop them ;
6. I plan to make a GUI (though not in Python, but in JS – I'm aware the webpatcher is eventually going to replace the Python GUI, and my Python knowledge is very limited – and I'll release everything I can under the MIT license just like the randomizer itself so when it'll be out, if you like it, feel free to integrate it into the webpatcher) ; for the groups, I'm probably going to extend my pattern matcher (that currently recognizes `!something`, `*something`, `something*` and combinations of the three) to treat `#something` as a group, that way we'll be able to write `#Junk` to target junk items, `#Bottle` to target bottles other than the call for help (for bottles including that one, there is already `Bottle*`), and maybe others – examples in 3. ;
7. As the examples given in 3. say, I'm thinking about supporting replacements as they are for people who want to start from a standard pool and add/remove items without doing anything in particular to them, an `extra` attribute in locations to let items that were already in the pool alone, and replace random junk with the item we're placing – and just fail the generation if there is no more junk to replace (I think the tool shouldn't make gameplay-altering decisions, such as to replace non-junk, by itself) ;
8. It seems to me that you're sometimes only considering the case where, as you say, the maker places everything by hand ; makers might also want to place only a few items to design part or all of a high-level playthrough and not care about the details, and therefore place some items but let randomness and logic work the rest out, for example : Morpha with Gold Scale and Iron Boots somewhere behind Morpha, and then Morpha's Boss Key somewhere out of the dungeon so that it's okay for the logic, but except for that specific cases, the maker might want to keep all keys in their respective dungeon and not care about their exact location, that would make keysanity not an option for them (another thing about plando + logic is that a novice maker might also want to place many things or even everything by hand, but still enable logic to have their placement sanity-checked).

**Comment by Exter-N** [comment]
9. For that case of placing things and asking logic for a sanity check, I'm considering adding a way to tell logic to ignore specific items (probably by assuming the player starts with it, with `state.collect`) while keeping the rest checked ; anything that might depend on ignored items would be on the maker's responsibility ;
10. One feature I'd want to add is to make the player really start with some items, but I noticed it would require patching the ROM, and I have pretty much no knowledge about that part. I would have to look for documentation about the save layout. What do you think about such a feature?

**Comment by Roman971** [comment]
I noticed something that I think should be addressed regarding this whole distribution file generation.
Originally there are 3 pieces of information that can be used to entirely generate a seed: the seed string, the setting string and the version string. With this plando implementation, there is a new piece of information needed: the distribution file data.
This means that it is no longer possible to generate a seed from just the seed, version and setting string like it's the case normally. The distribution data is also required now, and currently it's missing from the spoiler file, which makes it impossible to regenerate a seed from a spoiler file. However I think it's important, e.g. for debugging purposes, to keep it possible to regenerate the seed entirely from just the spoiler file content. 
Also a spoiler file corresponding to a seed generated from a distribution file might look weird or seem like something wrong happened, so additional information on the conditions of the generation must be available to avoid issues with bug reports.

So I think there needs to be a way for people to check, or at least recover, the distribution data from the spoiler file.
For that purpose I see 2 main possibilities, which could be used separately or together:

- Add all the distribution file data used for the seed generation to the spoiler file, probably at the bottom of the file in the original JSON format. This makes it easy to check directly which plando data was used during the generation.
- Generate a new "distribution" or "plando" string, unique for each distribution dataset and that could be easily encoded/decoded, and add it to the spoiler file. This makes it a copy-pasteable string and matches the existing format of seed/setting/version strings but would most likely result in very big strings due to how much information must be described, so it might not be a good idea.

Also, I found an edge bug related to generating seeds from a distribution file. When making the numeric seed, the raw output of the distribution file is used instead of the actual data, which means the seed will be different if irrelevant characters like whitespaces or carriage returns are added/removed in the json file, even if the actual data wasn't changed. To avoid this inconsistency, I think the json data should be parsed before being dumped into the numeric seed.

**Comment by Exter-N** [comment]
Closed in favor of https://github.com/TestRunnerSRL/OoT-Randomizer/tree/plando
